### PR TITLE
sc-transcation-pool refactor 

### PR DESCRIPTION
**Pull request by seunlanlege**
This is part of an ongoing client refactor.

- [x] Refactor all subsystems to use `TransactionPool` interface instead of concrete type `BasicPool`.
- [x] Move `TransactionPool` interface from `sp_transaction_pool` to `sc_transaction_pool_api`
- [x] Make `ChainApi` private as it doesn't need to be exposed to end users.
- [x] collapse `sc_transaction_graph` into `sc_transaction_pool`

polkadot companion: paritytech/polkadot#3419

### CODE REVIEW

File: client/consensus/manual-seal/src/lib.rs [hunk 1]
```diff
@@ -105,7 +105,7 @@ pub struct ManualSealParams<B: BlockT, BI, E, C: ProvideRuntimeApi<B>, A: ChainA
 	pub client: Arc<C>,
 
 	/// Shared reference to the transaction pool.
-	pub pool: Arc<Pool<A>>,
+	pub pool: Arc<P>,
```
Comment (gilescope) [inline client/consensus/manual-seal/src/lib.rs:108]: Maybe TP rather than just P?

File: client/transaction-pool/api/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+[package]
+name = "sc-transaction-pool-api"
+version = "3.0.0"
+authors = ["Parity Technologies <admin@parity.io>"]
+edition = "2018"
+license = "GPL-3.0-or-later WITH Classpath-exception-2.0"
+homepage = "https://substrate.dev"
+repository = "https://github.com/paritytech/substrate/"
+description = "Substrate transaction pool primitives."
```
Comment (tomusdrw) [inline client/transaction-pool/api/Cargo.toml:9]: Kind of confusing to have the same description in two crates now. Perhaps better to mention something about the API?

File: client/transaction-pool/api/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+readme = "README.md"
```
Comment (tomusdrw) [inline client/transaction-pool/api/Cargo.toml:10]: There is no readme.

File: client/transaction-pool/api/src/lib.rs [hunk 4]
```diff
@@ -17,6 +17,8 @@
 
 //! Transaction pool primitives types & Runtime API.
```
Comment (tomusdrw) [inline client/transaction-pool/api/src/lib.rs:18]: Outdated comment.

File: client/transaction-pool/api/src/lib.rs [hunk 5]
```diff
@@ -17,6 +17,8 @@
...
```
Comment (tomusdrw) [inline client/transaction-pool/api/src/lib.rs:19]:
```suggestion
#![warn(missing_docs)]
```

File: primitives/transaction-pool/src/lib.rs [hunk 6]
```diff
@@ -21,13 +21,6 @@
...
 pub mod runtime_api;
-#[cfg(feature = "std")]
-pub mod error;
-#[cfg(feature = "std")]
-mod pool;
-
-#[cfg(feature = "std")]
-pub use pool::*;
 
 pub use sp_runtime::transaction_validity::{
```
Comment (tomusdrw) [inline primitives/transaction-pool/src/lib.rs:25]: I guess it's not really needed that much any more?

File: primitives/transaction-pool/src/lib.rs [hunk 7]
```diff
@@ -21,13 +21,6 @@
 #![cfg_attr(not(feature = "std"), no_std)]
```
Comment (tomusdrw) [inline primitives/transaction-pool/src/lib.rs:21]: Please update module-level docs, cause it's just the runtime API now.

Perhaps worth to rename the crate?

File: primitives/transaction-pool/Cargo.toml [hunk 8]
```diff
@@ -14,26 +14,12 @@ readme = "README.md"
 targets = ["x86_64-unknown-linux-gnu"]
 
 [dependencies]
```
Comment (tomusdrw) [inline primitives/transaction-pool/Cargo.toml:16]: Update `description` please.

File: docs/Upgrading-2.0-to-3.0.md [hunk 9]
```diff
@@ -353,7 +353,7 @@ As said, we've added a new optional RPC service for improved light client suppor
 @@ -49,6 +49,7 @@ use sp_consensus::SelectChain;
  use sp_consensus_babe::BabeApi;
  use sc_rpc::SubscriptionTaskExecutor;
- use sp_transaction_pool::TransactionPool;
```
Comment (tomusdrw) [inline docs/Upgrading-2.0-to-3.0.md:356]: Isn't 3.0 already released? We probably shouldn't be updating that.

**Review by tomusdrw — COMMENTED** [review]
Not sure I fully get the purpose of moving this stuff around - I'm probably missing a larger context on what the refactoring is about, but left a few comments.

File: client/consensus/manual-seal/src/lib.rs [hunk 10]
```diff
@@ -94,7 +94,7 @@ pub fn import_queue<Block, Transaction>(
 }
 
 /// Params required to start the instant sealing authorship task.
-pub struct ManualSealParams<B: BlockT, BI, E, C: ProvideRuntimeApi<B>, A: ChainApi, SC, CS, CIDP> {
+pub struct ManualSealParams<B: BlockT, BI, E, C: ProvideRuntimeApi<B>, TP, SC, CS, CIDP> {
```
Comment (gilescope) [inline client/consensus/manual-seal/src/lib.rs:97]: Should TP have a trait bound here?

File: docs/Upgrading-2.0-to-3.0.md (hunk 9 above)
Comment (gilescope) [reply docs/Upgrading-2.0-to-3.0.md:356]: have a Upgrading-3.0-to-master.md?

File: primitives/transaction-pool/src/lib.rs (hunk 6 above)
Comment (gilescope) [reply primitives/transaction-pool/src/lib.rs:25]: I think as the primitives crate has got 'cleaner' a lot of these depednencies have moved up into the sc-transaction-pool crate.

File: client/consensus/manual-seal/src/lib.rs (hunk 10 above)
Comment (expenses) [reply client/consensus/manual-seal/src/lib.rs:97]: If it doesn't need a trait bound in the struct definition I think it's fine IMO. Sometimes it makes error messages clearer but in this case I wouldn't worry.

**Review by expenses — APPROVED** [review]
Seems pretty trivial, can't see anything wrong with it 👍🏻

File: client/consensus/manual-seal/src/lib.rs (hunk 10 above)
Comment (seunlanlege) [reply client/consensus/manual-seal/src/lib.rs:97]: yes it should, it's used here:
https://github.com/paritytech/substrate/blob/62939f986f44c638ac61a34438b8147685298fa5/client/consensus/manual-seal/src/lib.rs#L136

**Comment by seunlanlege** [comment]
Hmm so the companion build keeps failing, not sure why

**Comment by gilescope** [comment]
Am guessing that txpool_api needs to point to sc_ crate now not sp_ in `rpc/Cargo.toml` in polkadot

**Comment by seunlanlege** [comment]
cool so this just needs on more approval so i can forget about it

**Review by gilescope — APPROVED** [review]

**Comment by seunlanlege** [comment]
bot merge

**Pull request merged**